### PR TITLE
docs: playground loader

### LINKED ISSUE #581: Style and Design for Demo Playgrounds
@markcaron @marionnegp I did a quick pass at a design for these demos. If you read above, Benny wants to:

> show the snippet only first, with a button to load the demo playground

I like this approach as it saves a lot of vertical space. We can also use several web components to build this larger one.

https://xd.adobe.com/view/0779bcd8-06d0-49d7-86c3-de81534d70b1-522a/

cc @bennypowers

_Originally posted by @coreyvickery in https://github.com/RedHat-UX/red-hat-design-system/issues/467#issuecomment-1276672420_

### PR DESCRIPTION
**Pull request by bennypowers**
Closes #581
 
## What I did

1. removed the rhds bundle from the base template
    - if you want to load an element now, add it to the `importElements` list in frontmatter
    - all pages load cta and footer
    - element pages also load subnav, and their own element by tagname
1. fixed the cta shortcode when it contains markdown
1. Added a (docs-only) `<rh-playground>` custom element
    - it's got a button!
    - it's got a spinner!
    - it's got some animations and css filters!
    - it simplifies and encapsulates the demo loading code! 

## Testing Instructions
We're looking for undefined elements and missing icons
You can get a list of undefined elements like so:
```js
$$(':not(:defined)')
```
to see all the pf-icons on the page, use the find function in element inspector (ctrl-f) and type `pf-icon`

1. check the main page
2. check 'static' pages like `/foundations/`
3. check all the element pages

### CODE REVIEW

**Review by nikkimk — CHANGES_REQUESTED** [review]
### Site-wide
- `$$(':not(:defined)')` returns `[us-button]`
-  404 for `https://ux.redhat.com/assets/pfe.min.css`

### Accordion and Alert
- `$$(':not(:defined)')` returns `[playground-project, playground-tab-bar, playground-file-editor, playground-preview, us-button]`

### Avatar
- 404s for 
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/2p78mk1/demo/schoenfinkel.jpg`
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/2p78mk1/demo/hopper.jpg`

### Footer
- The following fail to load:
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/linkedin.js`
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/youtube.js`
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/facebook.js`
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/twitter.js`

### Secondary navigation
- `pf-icon` fails to load: `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/2sgu8rd/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/patternfly/arrow.js` 

### Tabs
- The following are errors:
  - `Uncaught TypeError: Failed to resolve module specifier "@patternfly/elements/pf-tabs/pf-tabs.js". Relative references must start with either "/", "./", or "../".`
   - `Uncaught Error: Unable to parse import map Unexpected end of JSON input in: at es-module-shims.js:866:17`
   - `Uncaught Error: Unable to resolve specifier '@patternfly/elements/pf-tabs/pf-tabs.js' imported from http://localhost:8081/elements/tabs/`
   - `Uncaught TypeError: Failed to resolve module specifier "/assets/rhds.min.js". Invalid relative url or base scheme isn't hierarchical.`

File: docs/elements/elements.html [hunk 1]
```diff
@@ -10,6 +10,11 @@
   - /assets/packages/@rhds/elements/elements/rh-subnav/rh-subnav-lightdom.css
 eleventyComputed:
   title: "{{ doc.pageTitle }} | {{ doc.slug | deslugify }}"
+  importElements:
+    - rh-subnav
+    - rh-cta
```
Comment (zeroedin) [inline docs/elements/elements.html:15]: If elements.html uses `layout-base.njk` which also  importsElements `rh-cta` and `rh-footer` do you need to repeat those here?  Assuming this is an override method not a incremental add.  Would the argument to do it this way be so you don't have to have to mental load of the full layout stack and which layouts import which elements?

**Review by nikkimk — CHANGES_REQUESTED** [review]
### All
- `$$(':not(:defined)')` returns `[us-button]`
- 404: `https://ux.redhat.com/assets/pfe.min.css`

## Accordion
-`$$(':not(:defined)')` returns `[playground-project, playground-tab-bar, playground-file-editor, playground-preview, us-button]`

## Alert
-`$$(':not(:defined)')` returns `[playground-project, playground-tab-bar, playground-file-editor, playground-preview, us-button]`

### Audio player
-`$$(':not(:defined)')` returns `[rh-subnav, rh-footer-universal, rh-footer-copyright, us-button]`
- `Uncaught TypeError: Failed to resolve module specifier "@rhds/elements/rh-audio-player/rh-audio-player.js". Relative references must start with either "/", "./", or "../".`

### Call to action
- `$$(':not(:defined)')` returns `[rh-subnav, rh-footer-universal, rh-footer-copyrigh, rh-alert, us-button]`
- `Uncaught ReferenceError: rh is not defined`
 
### Card 
- `$$(':not(:defined)')` returns `[rh-subnav, rh-footer-universal, rh-footer-copyright, us-button]`
- `Uncaught TypeError: Failed to resolve module specifier "@rhds/elements/rh-card/rh-card.js". Relative references must start with either "/", "./", or "../".`

### Footer
- `$$(':not(:defined)')` returns `[rh-alert, us-button]`
- The following fail to load:
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/linkedin.js`
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/youtube.js`
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/facebook.js`
   - `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/twitter.js`

### Jump links
- `$$(':not(:defined)')` returns `[rh-subnav, rh-footer-universal, rh-footer-copyright, us-button]`

### Primary navigation
- `$$(':not(:defined)')` returns `[rh-subnav, rh-cta, rh-alert, rh-footer-universal, rh-footer-copyright, us-button]`
- `Uncaught TypeError: Failed to resolve module specifier "@rhds/elements/rh-navigation/rh-navigation.js". Relative references must start with either "/", "./", or "../".`

### Secondary navigation
- `$$(':not(:defined)')` returns `[rh-alert, us-button]`
The following pf-icon fails to load: 
- Failed to load: `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/2sgu8rd/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/patternfly/arrow.js` 

### Popover
- `$$(':not(:defined)')` returns `[rh-subnav, rh-footer-universal, rh-footer-copyright, us-button]`
- `Uncaught TypeError: Failed to resolve module specifier "@rhds/elements/rh-popover/rh-popover.js". Relative references must start with either "/", "./", or "../".`

### Progress steps
- `$$(':not(:defined)')` returns `[rh-subnav, rh-footer-universal, rh-footer-copyright, us-button]`
- `Uncaught TypeError: Failed to resolve module specifier "@rhds/elements/rh-progress-steps/rh-progress-steps.js". Relative references must start with either "/", "./", or "../".`

### Tabs
- `Uncaught DOMException: Failed to execute 'define' on 'CustomElementRegistry': the name "pf-icon" has already been used with this registry`
 - `Uncaught DOMException: Failed to execute 'define' on 'CustomElementRegistry': the name “rh—buttonicon" has already been used with this registry`
 - `Uncaught DOMException: Failed to execute 'define' on 'CustomElementRegistry': the name "pf-tab” has already been used with this registry`

### Subnav
- `$$(':not(:defined)')` returns `[rh-subnav, rh-footer-universal, rh-footer-copyright, us-button]

File: docs/elements/elements.html (hunk 1 above)
Comment (bennypowers) [reply docs/elements/elements.html:15]: we can dedupe by routing through a `new Set()`, but ultimately

```
import './a.js';
import './a.js';
```
the browser will dedupe for you, so there's no correctness problem here ifaict.

**Comment by zeroedin** [comment]
I have a general concern about placing our demo html output front and center before loading the demo in the code block view.   

Specifically taking `navigation-secondary` into consideration as it is most likely the outlier but also shows the problem the most.  The default demo html has a lot of extra boiler plate to show off feature set of the component for the demo specifically which isn't meant to be copied and pasted by someone coming to try and use our components for the first time.

I personally would prefer a poster (for lack of better terms) set of html that represents maybe the default or bare minimum of the component load here before the "Load Demo" button is pushed?  Maybe also a view more examples link that links to the code tab for the component.   This is also a rhetorical question while I'm thinking about it and will start to explore that possibility myself as well.

**Comment by bennypowers** [comment]
> I have a general concern about placing our demo html output front and center before loading the demo in the code block view.
> 
> Specifically taking `navigation-secondary` into consideration as it is most likely the outlier but also shows the problem the most. The default demo html has a lot of extra boiler plate to show off feature set of the component for the demo specifically which isn't meant to be copied and pasted by someone coming to try and use our components for the first time.
> 
> I personally would prefer a poster (for lack of better terms) set of html that represents maybe the default or bare minimum of the component html to load here before the "Load Demo" button is pushed? Maybe also a link to view more examples that links to the code tab for the component. This is also a rhetorical question while I'm thinking about it and will start to explore that possibility myself as well.

so maybe we can simplify the main nav2 demo, and move the current demo to a "kitchen sink" file?

**Review by zeroedin — APPROVED** [review]
Looks good to me.  I did play around a bit with what it would be like if we added a fullscreen API functionality to the playground.  I do find it to be rather small and sometimes frustrating to look at when the demo code is long.  We can take a look at that as a future improvement.

**Comment by bennypowers** [comment]
@nikkimk Thank you for the careful review

> * `$$(':not(:defined)')` returns `[us-button]`
> * 404: `https://ux.redhat.com/assets/pfe.min.css`

Not related to this PR

> -`$$(':not(:defined)')` returns `[playground-project, playground-tab-bar, > playground-file-editor, playground-preview]`

Not a problem, as playground elements are lazyloaded

> -`$$(':not(:defined)')` returns `[rh-subnav, rh-footer-universal, > rh-footer-copyright]`

I checked a couple of the pages, and subnav, footer, etc were defined, so this 
looks like it was resolved. This will fail for pages belonging to elements that 
aren't yet defined. **I sent a commit to fix this, PTAL**.

> `@rhds/elements/rh-audio-player/rh-audio-player.js`
> `@rhds/elements/rh-card/rh-card.js`
> `@rhds/elements/rh-navigation/rh-navigation.js`
> `@rhds/elements/rh-popover/rh-popover.js`
> `@rhds/elements/rh-progress-steps/rh-progress-steps.js`

Expected, these elements are not yet merged. 

> ### Call to action
>     * `Uncaught ReferenceError: rh is not defined`

Couldn't 
[reproduce](https://deploy-preview-946--red-hat-design-system.netlify.app/elements/call-to-action/) 

> * `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/linkedin.js`
> * `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/youtube.js`
> * `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/facebook.js`
> * `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/23763is/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/fab/twitter.js`
> * `https://unpkg.com/playground-elements@0.17.0/__playground_swfs_1dae6563/2sgu8rd/node_modules/@patternfly/elements@2.0.5/pf-icon/icons/patternfly/arrow.js`


Confirmed, but unrelated to this PR. Icons are not loading in the playgrounds on 
prod as well, so this should be a new issue


> `Uncaught DOMException: Failed to execute 'define' on 'CustomElementRegistry':
> the name "pf-icon" has already been used with this registry`

> `Uncaught DOMException: Failed to execute 'define' on 'CustomElementRegistry':
> the name "rh—button" has already been used with this registry`

> `Uncaught DOMException: Failed to execute 'define' on 'CustomElementRegistry':
> the name "pf-tab" has already been used with this registry`

not sure what's causing this, but it doesn't break the page and the demos work.

**Comment by bennypowers** [comment]
@zeroedin playground now has a border and resize handle

**Pull request merged**